### ISSUE TITLE
Time-consuming gradient computation and backward for realizing meta network

### ISSUE THREAD

**Issue by Dataphz** [pytorch/pytorch#18682]
## ❓ Questions and Help
I am trying to implement the [paper](https://arxiv.org/pdf/1803.09050.pdf) for noisy label task. We found the training time is extremely time consuming, which troubled us for a long time.

The algorithm which i am tryinng to implement is the following:
![image](https://user-images.githubusercontent.com/30048368/55319105-c94bc080-54a6-11e9-923f-81b819c96154.png)

And the code is the following:
`meta_train_preds = meta_model(train_images)
            meta_train_loss = ((-1* torch.log(meta_train_preds + 1e-8).sum(dim=1)
                ).view(config.minibatch_size, config.group + 100 - 1).mean(dim=1))

            example_weights = torch.zeros(
                config.minibatch_size, requires_grad=True, dtype=torch.float32
            ).cuda()

            meta_train_weighted_loss = (example_weights * meta_train_loss).sum()

            meta_model.zero_grad()
            grads = torch.autograd.grad(
                meta_train_weighted_loss, (meta_model.params()), create_graph=True
            )
            meta_model.update_params(lr, source_params=grads)
            time_meta_train = time.time() - time_proc_data - time_iter_start

            time_start_val = time.time()
            meta_val_preds = meta_model(val_images)
            meta_val_loss = criterion(meta_val_preds, val_labels)
            time_val_forward = time.time() - time_start_val

            val_grad_to_examples_weight = torch.autograd.grad(
                meta_val_loss, example_weights, only_inputs=True
            )[0]

            time_val_backward = time.time() - time_start_val - time_val_forward

            clip_example_weights = torch.clamp(-val_grad_to_examples_weight, min=0)
            sum_weights = torch.sum(clip_example_weights)

            if sum_weights != 0:
                norm_example_weights = clip_example_weights / sum_weights
            else:
                norm_example_weights = clip_example_weights
            time_post_operation = (
                time.time() - time_start_val - time_val_forward - time_val_backward
            )
            time_meta_val = (
                time.time() - time_meta_train - time_proc_data - time_iter_start
            )`

We find time_backward in meta-val stage takes up 99% of the meta-val training time. Moreover, the meta-val training time takes up 93% of the all training time. 

Could anyone give some suggestions on how to resolve this issue? Thank you very much!

**Comment by soumith** [comment]
please use https://discuss.pytorch.org for questions
